source/core_mqtt.c

```diff
@@ -1814,20 +1814,27 @@ static size_t addEncodedStringToVector( uint8_t serailizedLength[ 2 ],
     const size_t seralizedLengthFieldSize = 2U;
     TransportOutVector_t * pLocalIterator = iterator;
     /* This function always adds 2 vectors. */
-    const size_t vectorsAdded = 2U;
+    size_t vectorsAdded = 0U;
 
     serailizedLength[ 0 ] = ( ( uint8_t ) ( ( length ) >> 8 ) );
     serailizedLength[ 1 ] = ( ( uint8_t ) ( ( length ) & 0x00ffU ) );
 
     /* Add the serialized length of the string first. */
     pLocalIterator[ 0 ].iov_base = serailizedLength;
     pLocalIterator[ 0 ].iov_len = seralizedLengthFieldSize;
+    vectorsAdded++;
+    packetLength = seralizedLengthFieldSize;
 
-    /* Then add the pointer to the string itself. */
-    pLocalIterator[ 1 ].iov_base = string;
-    pLocalIterator[ 1 ].iov_len = length;
-
-    packetLength = length + seralizedLengthFieldSize;
+    /* Sometimes the string can be NULL that is, of 0 length. In that case,
+     * only the length field should be encoded in the vector. */
+    if( ( string != NULL ) && ( length != 0U ) )
```

Comment:
What if string is null and length is non-zero? Should we just assert that string is non-null when length is nonzero?

Reply:
Oh true. Good catch

Reply:
Done!
That was a good catch.